### DISCUSSION TITLE
How to get LSP to work with `*-inl.h` files?

### DISCUSSION THREAD

**Discussion by nguyenvukhang** [facebook/CacheLib#344] [Q&A]
In the areas of the codebase I've explored, I observe a pattern of having some `*.h` file, and then an accompanying `*-inl.h` file. Take for example `cachelib/cachebench/cache/Cache.h` and `cachelib/cachebench/cache/Cache-inl.h`.

Among the `*.h` and `*-inl.h` pairs I've seen so far, I notice that the `*.h` file is the one and only file in the codebase that includes (`#include`) the corresponding `*-inl.h` file. However, LSP servers don't understand this. When I open a `*-inl.h` file, the LSP server is at a loss of where to find declarations when I, as a human, know to look in the corresponding `*.h` file.

My current fix, with the `clangd` lsp, is to include a flag that no one would use such as `__LSP_BE_HAPPY__` in a `.clangd` file at root:
```yml
CompileFlags:
  Add:
    - '-D__LSP_BE_HAPPY__'
```

and then manually make the following changes:
1. In `Cache.h`, make this edit:
   ```diff
   +#ifndef __LSP_BE_HAPPY__
    #include "cachelib/cachebench/cache/Cache-inl.h"
   +#endif
   ```
   To remove the recursion at LSP-analysis time.
2. In `Cache-inl.h`, prepend it with a
   ```c++
   #include "cachelib/cachebench/cache/Cache.h"
   ```
   So that the LSP server knows where to look for declarations or other `#include`s.

However, one issue remains. When I go to reference or go to definition in `Cache.h`, LSP doesn't know to search in `Cache-inl.h` (because it's no longer included).

So my question is this: is there an existing practice on getting LSP servers to understand the specific relationship between these two files?

**Comment by pbhandar2** [comment]
Closing this discussion as it has been inactive. If this is still relevant, then please create a new discussion post and we will get back to you as soon as we can. Thank you!
